pkg/storegateway/bucket.go

```diff
@@ -258,16 +259,38 @@ func (s *BucketStore) RemoveBlocksAndClose() error {
 }
 
 // Stats returns statistics about the BucketStore instance.
-func (s *BucketStore) Stats() BucketStoreStats {
+func (s *BucketStore) Stats(durations tsdb.DurationList) BucketStoreStats {
+	s.blocksMx.RLock()
+	defer s.blocksMx.RUnlock()
+
+	return buildStoreStats(durations, s.blocks)
+}
+
+func buildStoreStats(durations tsdb.DurationList, blocks map[ulid.ULID]*bucketBlock) BucketStoreStats {
 	stats := BucketStoreStats{}
+	stats.BlocksLoaded = make(map[time.Duration]int)
 
-	s.blocksMx.RLock()
-	stats.BlocksLoaded = len(s.blocks)
-	s.blocksMx.RUnlock()
+	if len(durations) != 0 {
+		for _, b := range blocks {
+			// Bucket each block into one of the possible block durations we're creating.
+			bucketed := bucketBlockDuration(durations, b.blockDuration())
+			stats.BlocksLoaded[bucketed]++
+		}
+	}
 
 	return stats
 }
 
+func bucketBlockDuration(buckets tsdb.DurationList, duration time.Duration) time.Duration {
+	for _, d := range buckets {
+		if duration <= d {
```

Comment:
Blocks are sometimes over by a few milliseconds which would put 2:00:00.1 blocks in the 12hr bucket, do we want to add a little jitter to this duration?

Reply:
After talking to @56quarters offline and inspecting some systems blocks that are MS over the time limit are very rare, so I don't think we need jitter here.

Reply:
Discussed offline. Based on [this code](https://github.com/prometheus/prometheus/blob/91054875d6465dd2dd1116f3f6c6446854fe16de/tsdb/compact.go#L309) I don't think it should be possible that blocks have more than 2h of samples. I've double checked a few production cells and couldn't find instances of blocks exceeding their supposed duration.

Reply:
> Blocks are sometimes over by a few milliseconds which would put 2:00:00.1 blocks in the 12hr bucket, do we want to add a little jitter to this duration?

Do we have blocks with time range **over** the expected one? I'm aware a block could be smaller (e.g. `1:59:59`) but not bigger. Can you follow up on this, please?